079-deprecation-and-removal-of-storage-overrides.md

```diff
@@ -0,0 +1,95 @@
+* Broker with node ID 2 uses the storage class `my-storage-class-zone-1c`
+* All other brokers use storage class `my-storage-class`
+
+Using a different storage class per broker can be useful in some special use-cases.
+For example when you need to guide the scheduling of the different brokers to different zones / nodes.
+But this kind of configuration has some challenges as well.
+Unless the broker ID is listed in the overrides, it will use the default storage class.
+So when you need to scale-up your Kafka cluster, you have to make sure it has the entries for the node IDs that will be added.
+
+## Motivation
+
+The overrides were very useful while all nodes were configured in the `Kafka` custom resource.
+But the Kafka Node Pools features was designed with the idea of using different configurations for different sets of nodes from the beginning.
+Users using node pools can have multiple different node pools - each representing a subset of Kafka nodes - with different configuration.
+This includes a different storage class.
+It also does need the user to specify exact node IDs as the storage configuration will be used for all nodes belonging to a given node pool.
+So it simplifies the configuration.
+
+In addition to simplifying the configuration for the users, it will also allow us to streamline the code and the API.
+In particular, the _storage diffing_ needed to avoid undesired changes will be simplified.
+
+## Proposal
+
+The persistent-volume-claim storage overrides will be deprecated immediately in Strimzi 0.43.
+The will remain supported and used, but users would be encouraged to move away from them and use Kafka Node Pools instead.
+This also falls into the time period when we expect most users to migrate to node pools and to KRaft which provides a good opportunity to move to per-node-pool storage class instead of using the overrides.
+
+Later, in the Strimzi version where support for ZooKeeper-based Kafka clusters is removed, the support for the storage overrides will be dropped and the overrides will be ignored.
+The overrides will remain in the CRD API for backwards compatibility but will not be used by the operator code.
+Finally, when moving the new `v1` CRD API in the future, the fields will be completely removed.
+
+Kubernetes do not allow changing storage class for existing persistent volumes. 
+If any user doesn't migrate from the overrides in time, the existing Kafka nodes and persistent volumes will not be affected.
+Only when the PVC/PV is deleted or when new nodes will be added during a scale-up, the overrides will be ignored and the default storage class will be used.
+So the impact on existing users should be minimal.
+
+### Warnings
+
+With the deprecation of the storage overrides, the operator will be updated to issue warnings when the overrides are used.
+The warnings will be printed in regular logs from the Cluster Operator.
+And they will be also added to the conditions in the `Kafka` CR status.
```

Comment:
Will we document how the users can get rid of these warnings specifically? Or do we expect users to figure out the equivalent node pool configuration for themselves?

Reply:
They get rid of them by stopping using the overrides. We already have content covering how to migrate to node pools and use them for different configurations. What exactly is the new configuration depends on the use-case. I.e. why exactly you use these overrides. For example, in the situations where I saw this used/recommended it to be used, it was used as a _trick_ to control the scheduling of individual brokers. In such case, the proper replacement might be different node pools with affinity rules rather than a different storage class. But obviously there might be also situations where the user actually needed different storage class. All of these cases can be covered by NodePools. But there is not a single exact _conversion_ procedure as it depends what do you do. Also, from my experience, this is very niche thing not used by every other user.

So as replied to Federico's comment - we can definitely have some bullet points to guide the users. But I do not see any need to have some elaborate multipage guide.

Reply:
> Or do we expect users to figure out the equivalent node pool configuration for themselves?

To also answer the question directly ... yes, I would expect the users to figure the right configuration by themselves. Given the expected number of people using it, I also do not expect it to be an issue if they come up asking for help.